iroh/src/magicsock/relay_actor.rs

```diff
@@ -346,6 +405,12 @@ impl RelayActor {
                     trace!("shutting down");
                     break;
                 }
+                Some(Err(err)) = self.active_relay_tasks.join_next() => {
```

Comment:
needs to check for `!.is_empty()`

Reply:
It (surprisingly) doesn't!  Which I also only realised once I thought I found a bug that this should cause a hot-loop but it didn't cause a hot-loop.

Because this `tokio::select!` call does not have an `else` branch the behaviour is as follows:
- the `self.active_relay_tasks.join_next()` future will will resolve immediately upon calling `tokio::select!`.
- The future output will not match the pattern `Some(Err(_))`.  **This disables the branch.**
- Because there are other enabled branches the select will continue waiting for the next future to resolve.
- Because there is not `else` branch, when all branches are disabled this will finally loop.
- But in any case, the last thing will never happen because one of the awaited futures is the cancellation token.

You could argue as to why rely on this behaviour?  But the branch disabling and `else` behaviour is exactly the same whether this branch is disabled by not matching the pattern or whether is is disabled by an if-clause.  So this doesn't change what you have to think about.  So it ends up with this only being a syntactic difference, not a difference in behaviour or cognitive load.  And the pattern-maching-disabling is less code and easier to read I think - once you understand it.

I've come to treat `tokio::select!` as the better enemy that I know.  So I am happy to rely fully on it for logic like this.

Reply:
in that case we should probably update the other places we do it like this, having this incosistent in the code base is going to confuse folks even more

Reply:
#3052